pkg/source/types.go

```diff
@@ -0,0 +1,117 @@
+package source
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+// nolint:revive
+type SourceType string
+
+const (
+	SourceTypeImage      SourceType = "image"
+	SourceTypeGit        SourceType = "git"
+	SourceTypeConfigMaps SourceType = "configMaps"
+	SourceTypeUpload     SourceType = "upload"
+	SourceTypeHTTP       SourceType = "http"
+)
+
+// TODO: Can we make source a bit more generic and composable/extensible?
```

Comment:
I was thinking something like:
```go
type Source struct {
    Type SourceType `json:"type"`
    Config json.RawMessage `json:"config"`
}
```
This would allow for different configurations to be parsed and validated by the individual unpacker types similar to how this was done for the [opm composite template in their respective builders](https://github.com/operator-framework/operator-registry/blob/6c602841934d6e154e38c0574cc140471dc063e6/alpha/template/composite/builder.go#L53-L78)

However, this ends up being a runtime validation rather than one that could occur on the CRD, *unless* we created a configurable webhook implementation that other projects could reuse for validating CRs with this source type before admitting it.

I'm wondering if the extensibility/configurability feature is valuable enough to warrant requiring users of the `source` package to implement an admission webhook if they want validation before their CR is created.

If we do decide it is worth it, I think this should be done in a follow up as there would likely be changes across the repo to support this and I feel like this PR is already relatively large